### PR TITLE
feat(manager): dynamic poetry install

### LINKED ISSUE #6095: Support dynamic python selection for poetry
**What would you like Renovate to be able to do?**

Detect required `python` and `poetry` versions

**Describe the solution you'd like**

Same as #6036 but also dynamically installing the right Poetry version

**Describe alternatives you've considered**

The current solution always runs latest Poetry

### PR DESCRIPTION
**Pull request by viceice**
Poetry will now be installed on the fly when using docker.

Closes #6095

### CODE REVIEW

File: lib/manager/poetry/artifacts.ts [hunk 1]
```diff
@@ -45,12 +72,18 @@ export async function updateArtifacts({
...
     }
+    const tagConstraint = getPythonConstraint(existingLockFileContent, config);
     const execOptions: ExecOptions = {
       cwdFile: packageFileName,
-      docker: { image: 'renovate/poetry' },
+      docker: {
+        image: 'renovate/python',
+        tagConstraint,
+        tagScheme: 'pep440',
+        preCommands: ['pip install poetry'],
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.ts:82]: Is there any way we can work out which version of poetry is in use?

Presumably there's at least a way to tell the difference between 0.x and 1.x so if it's only a few lines of work we should maybe do that at least.

File: lib/manager/poetry/__snapshots__/artifacts.spec.ts.snap [hunk 2]
```diff
@@ -59,19 +59,19 @@ Array [
 exports[`.updateArtifacts() returns updated poetry.lock using docker 1`] = `
 Array [
   Object {
-    "cmd": "docker pull renovate/poetry",
+    "cmd": "docker pull renovate/python:latest",
```
Comment (rarkins) [inline lib/manager/poetry/__snapshots__/artifacts.spec.ts.snap:62]: In the test it looks like python 3.7 should be used instead?

File: lib/manager/poetry/artifacts.spec.ts [hunk 3]
```diff
@@ -101,6 +103,7 @@ describe('.updateArtifacts()', () => {
         newPackageFileContent: '{}',
         config: {
           ...config,
+          compatibility: { python: '3.7' },
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.spec.ts:106]: Maybe we should have two docker tests - one for the autodetection of python version and another using the configured constraint.

File: lib/manager/poetry/artifacts.spec.ts [hunk 4]
```diff
@@ -70,7 +70,9 @@ describe('.updateArtifacts()', () => {
     expect(execSnapshots).toMatchSnapshot();
   });
   it('returns updated poetry.lock', async () => {
-    fs.readFile.mockResolvedValueOnce('Old poetry.lock' as any);
+    fs.readFile.mockResolvedValueOnce(
+      '[metadata]\npython-versions = ">3.6"' as never
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.spec.ts:74]: Here's a real example I found:

```
[metadata]
content-hash = "70609fddc0d3768b1003fc24207951ab7ad8bfad4c6cb326d6217c52f5a92e3d"
python-versions = "~2.7 || ^3.4"

```

File: lib/manager/poetry/artifacts.ts (hunk 1 above)
Comment (rarkins) [reply lib/manager/poetry/artifacts.ts:82]: Example: https://github.com/python-poetry/poetry/blob/9f93df1140efd44f37d07fc04fe5eb9abdc105db/pyproject.toml#L3

Comment (viceice) [reply lib/manager/poetry/artifacts.ts:82]: I think we need to read the `build-system.requires` section to get `poetry` constraint.

https://github.com/renovate-tests/poetry1/blob/4f41e4287d81d9ba3688f8a9cf7d3f01bb08d25b/pyproject.toml#L19

https://python-poetry.org/docs/pyproject/#poetry-and-pep-517

File: lib/manager/poetry/artifacts.spec.ts (hunk 4 above)
Comment (viceice) [reply lib/manager/poetry/artifacts.spec.ts:74]: mine is from https://github.com/renovate-tests/poetry1/blob/4f41e4287d81d9ba3688f8a9cf7d3f01bb08d25b/poetry.lock#L126

I've simply removed the hash, as it's not required for tests

File: lib/manager/poetry/extract.spec.ts [hunk 5]
```diff
@@ -56,7 +56,11 @@ describe('lib/manager/poetry/extract', () => {
     it('extracts multiple dependencies', () => {
       const res = extractPackageFile(pyproject1toml, filename);
       expect(res.deps).toMatchSnapshot();
-      expect(res.deps).toHaveLength(9);
+      expect(res.deps).toHaveLength(10);
+      expect(res.compatibility).toEqual({
+        poetry: 'poetry>=1.0 wheel',
```
Comment (rarkins) [inline lib/manager/poetry/extract.spec.ts:61]: Does this format work when put into `pip install` ?

Comment (viceice) [reply lib/manager/poetry/extract.spec.ts:61]: sure, tested locally:

image

File: lib/manager/poetry/__snapshots__/extract.spec.ts.snap [hunk 6]
```diff
@@ -115,6 +115,15 @@ Array [
       "nestedVersion": false,
     },
   },
+  Object {
+    "currentValue": "~2.7 || ^3.4",
```
Comment (rarkins) [inline lib/manager/poetry/__snapshots__/extract.spec.ts.snap:119]: Do we need to specify versioning=poetry here?

**Review by rarkins — DISMISSED** [review]

File: lib/manager/poetry/__snapshots__/extract.spec.ts.snap (hunk 6 above)
Comment (viceice) [reply lib/manager/poetry/__snapshots__/extract.spec.ts.snap:119]: maybe it makes sense to set `poetry` versioning as default at extract phase. I think this should be done in a seperate pr

Comment (rarkins) [reply lib/manager/poetry/__snapshots__/extract.spec.ts.snap:119]: ok

**Review by JamieMagee — DISMISSED** [review]

**Pull request merged**